### ISSUE TITLE
SplitLinePath takes a long time to render with certain curves

### ISSUE THREAD

**Issue by valtism** [airbnb/visx#1591]
For some reason, `<SplitLinePath>` is taking a long time to render when paired with certain curves.

**Reproduction:** 

https://stackblitz.com/edit/react-ts-mgzfrj?file=App.tsx 

Open the above link in a new tab and react profiling tools and you will see that the render is taking a long time (>100ms on a new M1 Mac)

![Screen Shot 2022-10-25 at 12 22 19](https://user-images.githubusercontent.com/1286001/197863380-a6686fcc-e8bf-4422-9597-8c54fdedbac2.png)

This only seems to occur for smooth curves like `curveMonotoneX`, `curveNatural` and `curveCardinal`.

This only seems to affect `<SplitLinePath>` while rendering the same data with the same curves as two separate line paths only takes a couple of milliseconds.

**Comment by williaster** [comment]
Hey @valtism thanks for reporting this 🙏 my guess would be the algorithm it's using to sample points for some curves might not be performant. @kristw any thoughts?

**Comment by valtism** [comment]
@williaster So after some investigation it looks like most of the time is being spent in `pathElement.getPointAtLength(distance)` https://github.com/airbnb/visx/blob/master/packages/visx-shape/src/util/getSplitLineSegments.ts#L74

This is a very expensive call (~0.2ms-0.4ms) that is called in a `for` loop around 400 times for a path like the above (~120ms all up). I remember running into performance issues with it using it for Drag https://github.com/airbnb/visx/blob/master/packages/visx-drag/src/util/useSamplesAlongPath.ts

I am not sure what fix would be right here but if we can avoid using `getPointAtLength` we would greatly speed up this function.

**Comment by valtism** [comment]
So I was wondering; since we already have the path, can't we chop the path up at each point the segment changes? We need to combine them at first to ensure a smooth curve the whole way, but after that we can use our segment data to find the first points at each segment and use that to cut the path into segments. I'm not yet sure if this will work with all curves (I think some curves have the line miss the actual data point) but I have a working demo for our use-case with `curveMonotoneX`: https://gist.github.com/valtism/7af5334243ee6a812f6d890f5a52092a

This case seems less flexible than the current approach, but it is ~500x faster in this example.

**Comment by kristw** [comment]
Another thing that might help is rounding the `x` and `y` position to fewer decimal points as svg can be much slower due on the length of path string.

```tsx
<SplitLinePath x={x => x.toFixed(2)} y={y => y.toFixed(2)} />
```

**Comment by valtism** [comment]
I have tried adjusting the sample rate. At a rate of 2 I would say you get a mostly acceptable quality, but there is noticeable degradation of the curve. Anything more seems unacceptably degraded. Although a rate of 2 would halve the calculation time, it is still significant. 

The rounding idea is interesting. I don't think it will have a huge impact but let me test it out quickly now

**Comment by valtism** [comment]
I can confirm that rounding the `x` and `y` position decimals makes little to no performance difference.

**Comment by kristw** [comment]
Thanks for trying. 

> I'm not yet sure if this will work with all curves (I think some curves have the line miss the actual data point) but I have a working demo for our use-case with curveMonotoneX: https://gist.github.com/valtism/7af5334243ee6a812f6d890f5a52092a
> This way seems less flexible than the current approach, and certainly is a breaking change, but it is ~500x faster in my use-case.

Thanks for sharing the idea. I would love to improve the perf but am reluctant to introduce breaking changes which will make it not compatible with other interpolations. Perhaps you can use the custom component as the workaround for your case for now.

**Comment by valtism** [comment]
Yeah, that's the plan. Already implementing it in our codebase right now. I can share the end-result when I'm done. It could be exposed as a new shape, like `<PerformantSplitLine>` or something

**Comment by valtism** [comment]
So this is a rough solution I have to meet our use-case: https://gist.github.com/valtism/5a6265b503aa7a7ae2a765f0d11163cf

One of the main differences is that this approach no longer relies on exact matching of points (which is impossible with some curves) and instead uses the `getClosest` function to find the closest `x` point in the path to the target. It now mostly works for all types of curves (although not yet for closed curves).

This was only set up for `x` segmentation, but could be extended fairly easily to support both `x` and `y`. I haven't though about how `length` would work, because we really only need `x` segmentation.

**Comment by valtism** [comment]
Revisited this when working on a different project, and noticed that this can be waaaay way simplified. Don't know why I made it so complex in the first place ¯\_(ツ)_/¯ 

Anyway, here's what you need for your own Performant `SplitLinePath`: https://gist.github.com/valtism/80d8d7d6f3ae58c29d274210bda7990c

**Comment by hshoff** [comment]
@valtism nice, looks good to me.

We could add a `useSegementedLinePath()` hook that does the calculations and returns the `paths` and could be used with a `<LinePath>`.

Then we could also add a `<SegementedLinePath>` component for the drop-in replacement of `<SplitLinePath>`. Would be an opt-in change vs breaking change.

**Comment by valtism** [comment]
Either of these seem like good solutions to me 👍

**Comment by Demnu** [comment]
I've been running into the same problem I'm going to give @valtism's solution a go. For anyone else interested, here is an example that shows the performance difference between SplitLinePath and LinePath.

[LinePath vs SplitLinePath](https://codesandbox.io/p/sandbox/romantic-morse-kr6syq)
